File: Design and programming environments/Iteration 1/TicketManager/src/main/java/Repository/EventRepository.java
package Repository;

import Domain.Event;
import Utils.JdbcUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import javax.sound.midi.Soundbank;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class EventRepository implements IRepository<Integer, Event> {

    private JdbcUtils dbUtils;

    private static final Logger logger= LogManager.getLogger();

    public EventRepository(Properties props){
        logger.info("Initializing EventRepository with properties: {} ",props);
        dbUtils=new JdbcUtils(props);
    }

    @Override
    public int size() {
        logger.traceEntry();
        Connection con=dbUtils.getConnection();

        try(PreparedStatement preStmt=con.prepareStatement("select count(*) as [SIZE] from Events")){
            try(ResultSet result = preStmt.executeQuery()){
                if(result.next()){
                    logger.traceExit(result.getInt("SIZE"));
                    return result.getInt("SIZE");
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB Event Size " + e);
        }
        return 0;
    }

    @Override
    public void save(Event entity) {
        logger.traceEntry("saving event{}",entity);
        Connection con=dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("insert into Events values (?,?,?)")) {
            preStmt.setInt(1,entity.getId());
            preStmt.setString(2,entity.getName());
            preStmt.setInt(3,entity.getSeats());
            int result = preStmt.executeUpdate();
        }
        catch (SQLException e) {
            logger.error(e);
            System.out.println("Error DB Event Save "+e);
        }
        logger.traceExit();
    }

    @Override
    public void delete(Integer integer) {
        logger.traceEntry("deleting event with {}",integer);
        Connection con = dbUtils.getConnection();
        try(PreparedStatement preStmt=con.prepareStatement("delete from Events where id=?")){
            preStmt.setInt(1,integer);
            int result = preStmt.executeUpdate();
        }catch(SQLException e){
            logger.error(e);
            System.out.println("Error DB Event Delete "+e);
        }
        logger.traceExit();
    }

    @Override
    public void update(Integer integer, Event entity) {
        logger.traceEntry("updating event with {}",integer);
        Connection con = dbUtils.getConnection();
        Event event = findOne(integer);
        if(event==null){
            return;
        }
        Integer newId=entity.getId();
        String newName=entity.getName();
        Integer newSeats=entity.getSeats();

        try(PreparedStatement preStmt=con.prepareStatement("update Events"+
                " set id = ?,name = ?,seats = ? "+
                " where id = ? ")){
            preStmt.setInt(4,integer);
            preStmt.setInt(1,newId);
            preStmt.setString(2,newName);
            preStmt.setInt(3,newSeats);
            int result=preStmt.executeUpdate();
        }catch(SQLException e) {
            logger.error(e);
            System.out.println("Error DB Event Update " + e);
        }
        logger.traceExit();
    }

    @Override
    public Event findOne(Integer integer) {
        logger.traceEntry("finding event with id {}",integer);
        Connection con=dbUtils.getConnection();

        try(PreparedStatement preStmt=con.prepareStatement("select * from Events where id=?")){
            preStmt.setInt(1,integer);
            try(ResultSet result = preStmt.executeQuery()){
                if(result.next()){
                    int id= result.getInt("id");
                    String name=result.getString("name");
                    int seats=result.getInt("seats");
                    Event event = new Event(id,name,seats);
                    logger.traceExit(event);
                    return event;
                }
            }
        }catch(SQLException e) {
            logger.error(e);
            System.out.println("Error DB Event FindOne " + e);
        }
        logger.traceExit("No event found with id{}",integer);
        return null;
    }

    @Override
    public Iterable<Event> findAll() {
        logger.traceEntry();
        Connection con=dbUtils.getConnection();
        List<Event> events=new ArrayList<>();
        try(PreparedStatement preStmt=con.prepareStatement("select * from events")){
            try(ResultSet result=preStmt.executeQuery()){
                while(result.next()){
                    int id= result.getInt("id");
                    String name=result.getString("name");
                    int seats=result.getInt("seats");
                    Event event = new Event(id,name,seats);
                    events.add(event);
                }
            }
        }catch(SQLException e) {
            logger.error(e);
            System.out.println("Error DB Event FindAll " + e);
        }
        logger.traceExit(events);
        return events;
    }
}

File: Design and Programming Environments/Iteration 3/TicketManagerCSharp/TicketManagerCSharp/repository/EventRepository.cs
using log4net;
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Text;
using TicketManagerCSharp.domain;
using TicketManagerCSharp.utils;

namespace TicketManagerCSharp.repository
{
    class EventRepository : IRepository<int, Event>
    {
        private static readonly ILog log = LogManager.GetLogger("EventRepository");

        IDictionary<String, string> props;

        public EventRepository(IDictionary<String,string> props)
        {
            log.Info("Creating EventRepository...");
            this.props = props;
        }

        public void Delete(int id)
        {
            log.InfoFormat("Entering Delete with new value {0}...", id);
            IDbConnection con = DBUtils.getConnection(props);

            using(var comm = con.CreateCommand())
            {
                comm.CommandText = "delete from Events where id=@id";
                IDbDataParameter idParam = comm.CreateParameter();
                idParam.ParameterName = "@id";
                idParam.Value = id;
                comm.Parameters.Add(idParam);
                var rez = comm.ExecuteNonQuery();
                if (rez == 0)
                {
                    throw new RepositoryException("Error:Can't delete Event!");
                }
                log.InfoFormat("Exiting Delete");
            }
        }

        public IEnumerable<Event> FindAll()
        {
            IDbConnection con = DBUtils.getConnection(props);
            log.InfoFormat("Entering findAll...");
            IList<Event> events = new List<Event>();
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "select id,name,seats from Events";
                using(var dataR = comm.ExecuteReader())
                {
                    while (dataR.Read())
                    {
                        int idE = dataR.GetInt32(0);
                        string name = dataR.GetString(1);
                        int seats = dataR.GetInt32(2);
                        Event e = new Event(idE, name, seats);
                        events.Add(e);
                    }
                }
            }
            log.InfoFormat("Exiting findAll");
            return events;
        }

        public Event FindOne(int id)
        {
            log.InfoFormat("Entering findOne with value {0}", id);
            IDbConnection con = DBUtils.getConnection(props);

            using(var comm = con.CreateCommand())
            {
                comm.CommandText = "select id,name,seats from Events where id=@id";
                IDbDataParameter paramId = comm.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                comm.Parameters.Add(paramId);

                using(var dataR= comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        int idE = dataR.GetInt32(0);
                        string name = dataR.GetString(1);
                        int seats = dataR.GetInt32(2);
                        Event e = new Event(idE, name, seats);
                        log.InfoFormat("Exiting findOne with value{0}", e);
                        return e;
                    }
                }
            }
            log.InfoFormat("Exiting findOne with value {0}", null);
            return null;
        }

        public void Save(Event entity)
        {
            log.InfoFormat("Entering Save with new value {0}...", entity);
            var con = DBUtils.getConnection(props);
            using (var comm = con.CreateCommand())
            {
                comm.CommandText = "insert into Events(name,seats) values (@name,@seats)";
                var nameParam = comm.CreateParameter();
                nameParam.ParameterName = "@name";
                nameParam.Value = entity.Name;
                comm.Parameters.Add(nameParam);

                var seatsParam = comm.CreateParameter();
                seatsParam.ParameterName = "@seats";
                seatsParam.Value = entity.Seats;
                comm.Parameters.Add(seatsParam);

                var result = comm.ExecuteNonQuery();
                if (result == 0)
                {
                    throw new RepositoryException("Error:Can't add event!");
                }
                log.InfoFormat("Exiting Save");
            }
        }

        public int Size()
        {
            log.InfoFormat("Calculating Events database Size...");
            IDbConnection con = DBUtils.getConnection(props);
            int size = -1;
            using(var comm = con.CreateCommand())
            {
                comm.CommandText = "select count(*) from Events";

                var rez = comm.ExecuteNonQuery();
                if (rez == 0)
                {
                    throw new RepositoryException("Error:Can't calculate size!");
                }

                using(var dataR= comm.ExecuteReader())
                {
                    if (dataR.Read())
                    {
                        size = dataR.GetInt32(0);
                    }
                }
            }
            log.InfoFormat("Finished calculation!");
            return size;
        }

        public void Update(int id, Event entity)
        {
            log.InfoFormat("Entering Update with value {0}", id);
            var con = DBUtils.getConnection(props);

            using (var comm= con.CreateCommand())
            {
                comm.CommandText = "update Events set name=@name,seats=@seats where id=@id";
                var idParam = comm.CreateParameter();
                idParam.ParameterName = "@id";
                idParam.Value = entity.Id;
                comm.Parameters.Add(idParam);

                var nameParam = comm.CreateParameter();
                nameParam.ParameterName = "@name";
                nameParam.Value = entity.Name;
                comm.Parameters.Add(nameParam);

                var seatsParam = comm.CreateParameter();
                seatsParam.ParameterName = "@seats";
                seatsParam.Value = entity.Seats;
                comm.Parameters.Add(seatsParam);

                var result = comm.ExecuteNonQuery();
                if (result == 0)
                {
                    throw new RepositoryException("Error:Can't update Event!");
                }
                log.InfoFormat("Exiting Update");
            }
        }
    }
}
